src/atama/head/head.cpp

```diff
@@ -0,0 +1,576 @@
+  object_count = 0;
+
+  joints = {};
+  min_time = -1;
+}
+
+bool Head::init_scanning()
+{
+  if (scan_init) {
+    return false;
+  }
+
+  return scan_init = true;
+}
+
+void Head::initialize()
+{
+  pan_angle = current_pan_angle;
+  tilt_angle = -current_tilt_angle;
+
+  pan_error = 0;
+  pan_error_difference = 0;
+
+  tilt_error = 0;
+  tilt_error_difference = 0;
+}
+
+void Head::move_by_angle(double pan_angle, double tilt_angle)
+{
+  function_id = Head::MOVE_BY_ANGLE;
+  stop_scan();
+  this->pan_angle = pan_center + keisan::clamp(pan_angle, right_limit, left_limit);
+  this->tilt_angle = tilt_center + keisan::clamp(tilt_angle, bottom_limit, top_limit);
+}
+
+void Head::tracking(double pan, double tilt)
+{
+  pan_error_difference = pan - pan_error;
+  pan_error = pan;
+
+  tilt_error_difference = tilt - tilt_error;
+  tilt_error = tilt;
+
+  pan_only = false;
+  tilt_only = false;
+
+  tracking();
+}
+
+void Head::tracking_pan_only(double pan)
+{
+  pan_error_difference = pan - pan_error;
+  pan_error = pan;
+
+  pan_only = true;
+
+  tracking();
+}
+
+void Head::tracking_tilt_only(double tilt)
+{
+  tilt_error_difference = tilt - tilt_error;
+  tilt_error = tilt;
+
+  tilt_only = true;
+
+  tracking();
+}
+
+void Head::tracking()
+{
+  stop_scan();
+
+  if (tilt_only) {
+    pan_angle = current_pan_angle;
+  } else {
+    double p_offset = pan_error * pan_p_gain;
+    p_offset *= p_offset;
+    if (pan_error < 0) {
+      p_offset = -p_offset;
+    }
+
+    double d_offset = pan_error_difference * pan_d_gain;
+    d_offset *= d_offset;
+    if (pan_error_difference < 0) {
+      d_offset = -d_offset;
+    }
+
+    pan_angle += (p_offset + d_offset);
+  }
+
+  if (pan_only) {
+    tilt_angle = current_tilt_angle;
+  } else {
+    double p_offset = tilt_error * tilt_p_gain;
+    p_offset *= p_offset;
+    if (tilt_error < 0) {
+      p_offset = -p_offset;
+    }
+
+    double d_offset = tilt_error_difference * tilt_d_gain;
+    d_offset *= d_offset;
+    if (tilt_error_difference < 0) {
+      d_offset = -d_offset;
+    }
+    tilt_angle += (p_offset + d_offset);
+  }
+
+  pan_angle = pan_center + keisan::clamp(pan_angle - pan_center, right_limit, left_limit);
+  tilt_angle = tilt_center + keisan::clamp(tilt_angle - tilt_center, bottom_limit, top_limit);
+
+  // set calculation result
+  if (!joints.empty()) {
+    joints[0].set_position(pan_angle);
+    joints[1].set_position(tilt_angle);
+  }
+}
+
+void Head::scan(int mode)
+{
+  start_scan();
+
+  if (scan_mode == mode) {
+    return;
+  }
+
+  scan_mode = mode;
+  scan_init = false;
+}
+
+void Head::set_scan_limit(
+  double left_limit, double right_limit,
+  double top_limit, double bottom_limit)
+{
+  scan_left_limit = left_limit;
+  scan_right_limit = right_limit;
+  scan_top_limit = top_limit;
+  scan_bottom_limit = bottom_limit;
+}
+
+void Head::scan_process()
+{
+  for (int i = 0; i < static_cast<int>(joints.size()); i++) {
+    joints[i].set_pid_gain(20.0, 3.0, 2.0);
+  }
```

Comment:
i suggest to remove `static_cast` here, use `size_t` instead of `int`, and use `++i` instead of `i++`.